src/mgr/ReportStream.h

```diff
@@ -0,0 +1,95 @@
+/*
+ * Copyright (C) 1996-2023 The Squid Software Foundation and contributors
+ *
+ * Squid software is distributed under GPLv2+ license and includes
+ * contributions from numerous individuals and organizations.
+ * Please see the COPYING and CONTRIBUTORS files for details.
+ */
+
+/* DEBUG: section 16    Cache Manager API */
+
+#ifndef SQUID__SRC_MGR_REPORTSTREAM_H
+#define SQUID__SRC_MGR_REPORTSTREAM_H
+
+#include "base/PackableStream.h"
+#include "mgr/forward.h"
+
+namespace Mgr
+{
+
+/// interface for Cache Manager report formatters
+class ReportStream : public RefCountable, public PackableStream
+{
+public:
+    ReportStream(Packable &buf) : PackableStream(buf) {}
+
+    virtual ReportStream &reportStart(ReportStream &os) { return os; }
+    virtual ReportStream &reportEnd(ReportStream &os) { os << std::endl; return os; }
+
+    virtual ReportStream &tableStart(ReportStream &os) { return os; }
+    virtual ReportStream &tableEnd(ReportStream &os) { firstRowDone = false; firstCellDone = false; return os; }
+    virtual ReportStream &tableRowStart(ReportStream &os) { firstRowDone = true; return os; }
+    virtual ReportStream &tableRowEnd(ReportStream &os) { firstCellDone = false; os << std::endl; return os; }
+    virtual ReportStream &tableCellStart(ReportStream &os) { firstCellDone = true; return os; }
+    virtual ReportStream &tableCellEnd(ReportStream &os) { return os; }
+protected:
+    bool firstRowDone = false;
+    bool firstCellDone = false;
+};
+
+/// std::ostream with text/plain formatting for Cache Manager reports
+class ReportPlain : public ReportStream
+{
+public:
+    ReportPlain(Packable &buf) : ReportStream(buf) {}
+
+    /* Mgr::ReportStream API */
+    ReportStream &tableRowStart(ReportStream &os) override {
+        os << ' ';
+        return ReportStream::tableRowStart(os);
+    }
+    ReportStream &tableCellStart(ReportStream &os) override {
+        if (firstCellDone)
+            os << '\t';
+        return ReportStream::tableCellStart(os);
+    }
+};
+
+/// std::ostream with text/yaml formatting for Cache Manager reports
+class ReportYaml : public ReportStream
+{
+public:
+    ReportYaml(Packable &buf) : ReportStream(buf) {}
+
+    /* Mgr::ReportStream API */
+    ReportStream &tableRowStart(ReportStream &os) override {
+        os << " - [";
+        return ReportStream::tableRowStart(os);
+    }
+    ReportStream &tableRowEnd(ReportStream &os) override {
+        os << ']';
+        return ReportStream::tableRowEnd(os);
+    }
+    ReportStream &tableCellStart(ReportStream &os) override {
+        if (firstCellDone)
+            os << ", ";
+        return ReportStream::tableCellStart(os);
+    }
+};
+
+} // namespace Mgr
+
+#define MGR_IOMANIP(SYMBOL,METHOD) std::ostream &SYMBOL(std::ostream &os) { \
+    if (auto s = dynamic_cast<Mgr::ReportStream*>(&os)) s->METHOD(*s); \
```

Comment:
> > PR description: Adds custom stream class for display formatting Mgr::Action reports and I/O manipulators for use by report generators to embed syntax markup without knowing the specific syntax tokens.
> 
> ```c++
> class ReportPlain : public ReportStream
> ...
> class ReportYaml : public ReportStream
> ```
> 
> This PR proposes to support dynamic (i.e. at runtime) selection of cache manager report formats. I am not convinced the extra complexity and dynamic_cast hacks associated with that feature are worth its benefits.

This PR is following your requirements for `ReportFormater` being a stream object - the casting is a side effect of that requirement imposed by `std::ostream` API.

You may recall my original design plan was to have `FormatFoo` objects whose virtual methods produce the relevant `Foo` syntax token or delimiter as a string. The `Action` producing a report would pass that opaque token to the stream. @kinkie is currently also leaning that way in recent PRs with the `spaces()` API for handling YAML indentation.


> Moreover, I doubt that static (i.e. at compile time) support for multiple output formats is worth supporting long-term. I believe we picked YAML format because, in part, we could convert _all_ cache manager reports to that format, without sacrificing reports readability by humans while improving their consumption by programs. I think we should stick with that simple plan.

I for one, picked YAML format because many of the semi-structured reports are/were already nearly-valid YAML and it is best practice not to re-invent formats if a perfectly suitable one already exists. The alternative was JSON which is a more limited sub-set of YAML, while popular would require more syntax fluff to be added.

IME "all" is going a bit too far. There are 4 distinct formats that are suitable, depending on which "report" being output:
 1. `text/html` (`mgr:index`) - for Browser display
 2. `text/plain` (the BasicActions responses eg shutdown/reconfigure/restart) - for WebUI or Browser display
 3. `text/tsv` (tabular reports eg. mem/histograms/io/events) - for import directly by spreadsheet applications
 4. `text/yaml` (reports with complex/structured content) - for use by any YAML/JSON capable tools

Technically we could merge (3) and (4), but the tables are most useful being able to manipulate as spreadsheets without converting from YAML/JSON.

We could also merge (1) and (2) as HTML. But that raises issues about compatibility with scripts and command line tools that may be expecting the existing plain-text string.

Reply:
I had forgotten about this work, thanks for bubbling it up!
I would do away with the TSV reporting; the work I'm doing is proving, I think, that we can do without it, and leave it to postprocessing to handle it. Doing so would simplify the API greatly. spaces() could be folded in work. Leaving reports aware of their output format would make them less generic but would IMO allow for lower implementation effort.

This said, I wouldn't block the reformatting work I'm doing while waiting for this to PR to land, at the cost of me redoing the work. We are currently blocked on multiple streams while deciding which way to go, and I'd rather move in the right direction even if not to the final destination than standing still

Reply:
> > > PR description: Adds custom stream class for display formatting Mgr::Action reports and I/O manipulators for use by report generators to embed syntax markup without knowing the specific syntax tokens.
> > 
> > ```c++
> > class ReportPlain : public ReportStream
> > ...
> > class ReportYaml : public ReportStream
> > ```
> >   
> > Alex: This PR proposes to support dynamic (i.e. at runtime) selection of cache manager report formats. I am not convinced the extra complexity and dynamic_cast hacks associated with that feature are worth its benefits.
> 
> This PR is following your requirements for `ReportFormater` being a stream object - the casting is a side effect of that requirement imposed by `std::ostream` API.

I believe the problems identified in this change request can and should be addressed. I do not know what "your requirements" you are referring to, so I do not know whether they were wrong or orthogonal to this change request. Please provide a reference.

> You may recall my original design plan was to have `FormatFoo` objects whose virtual methods produce the relevant `Foo` syntax token or delimiter as a string. The `Action` producing a report would pass that opaque token to the stream.

I do not recall that, but the above summary does not sound like something we should adopt.

> @kinkie is currently also leaning that way in recent PRs with the `spaces()` API for handling YAML indentation.

FWIW, I see no relevant connection between a trivial function producing a given number of space characters and "objects whose virtual methods produce the relevant `Foo` syntax token".

> > Alex: Moreover, I doubt that static (i.e. at compile time) support for multiple output formats is worth supporting long-term. I believe we picked YAML format because, in part, we could convert _all_ cache manager reports to that format, without sacrificing reports readability by humans while improving their consumption by programs. I think we should stick with that simple plan.

> IME "all" is going a bit too far. There are 4 distinct formats that are suitable, depending on which "report" being output:

I hope that a single format (YAML) is also "suitable" and, hence, should be used as the only format (at least for the foreseeable future), simplifying things.

> 1. `text/html` (`mgr:index`) - for Browser display

IMO, we should (remove `mgr:index` support and) not target browsers as cache manager report consumers. Squid management via nice web pages should not be Squid responsibility in the foreseeable future.

> 2. `text/plain` (the BasicActions responses eg shutdown/reconfigure/restart) - for WebUI or Browser display

YAML works just fine for these AFAICT. These trivial responses do not need to be rendered by browsers, but if somebody wants them rendered in a browser, they will be rendered OK.

> 3. `text/tsv` (tabular reports eg. mem/histograms/io/events) - for import directly by spreadsheet applications

I suspect that (appropriately formatted) YAML report can be imported into a spreadsheet without a lot of effort. The utility of such imports is too low to justify adding another format IMO.
 
> 4. `text/yaml` (reports with complex/structured content) - for use by any YAML/JSON capable tools

... and humans.
 

> Technically we could merge (3) and (4), but the tables are most useful being able to manipulate as spreadsheets without converting from YAML/JSON.

I disagree that those tables are "most useful being able to manipulate as spreadsheets", but I hope they can be YAML-formatted in such a way that they can be imported into spreadsheets without a lot of effort.